File: Session_11_Backtracking/multi_language_implementations/n_queens.java
// Session 11: N-Queens Backtracking in Java

public class n_queens {
    private static final int N = 4;

    public static void printBoard(int[][] board) {
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                System.out.print(board[i][j] != 0 ? " Q " : " . ");
            }
            System.out.println();
        }
        System.out.println();
    }

    public static boolean isSafe(int[][] board, int row, int col) {
        int i, j;
        for (i = 0; i < col; i++) {
            if (board[row][i] != 0) {
                return false;
            }
        }
        for (i = row, j = col; i >= 0 && j >= 0; i--, j--) {
            if (board[i][j] != 0) {
                return false;
            }
        }
        for (i = row, j = col; j >= 0 && i < N; i++, j--) {
            if (board[i][j] != 0) {
                return false;
            }
        }
        return true;
    }

    public static boolean solveNQueensHelper(int[][] board, int col) {
        if (col >= N) {
            return true;
        }
        for (int i = 0; i < N; i++) {
            if (isSafe(board, i, col)) {
                board[i][col] = 1;
                if (solveNQueensHelper(board, col + 1)) {
                    return true;
                }
                board[i][col] = 0; // backtrack
            }
        }
        return false;
    }

    public static void main(String[] args) {
        int[][] board = new int[N][N];
        if (solveNQueensHelper(board, 0)) {
            printBoard(board);
        } else {
            System.out.println("No solution");
        }
    }
}

File: Session_11_Backtracking/multi_language_implementations/n_queens.py
# Session 11: N-Queens Backtracking in Python

N = 4

def print_board(board):
    for i in range(N):
        for j in range(N):
            print(" Q " if board[i][j] else " . ", end="")
        print()
    print()

def is_safe(board, row, col):
    # Check this row on left side
    for i in range(col):
        if board[row][i]:
            return False
            
    # Check upper diagonal on left side
    i, j = row, col
    while i >= 0 and j >= 0:
        if board[i][j]:
            return False
        i -= 1
        j -= 1
        
    # Check lower diagonal on left side
    i, j = row, col
    while j >= 0 and i < N:
        if board[i][j]:
            return False
        i += 1
        j -= 1
        
    return True

def solve_n_queens_helper(board, col):
    if col >= N:
        return True
        
    for i in range(N):
        if is_safe(board, i, col):
            board[i][col] = 1
            if solve_n_queens_helper(board, col + 1):
                return True
            board[i][col] = 0  # backtrack
            
    return False

def main():
    board = [[0] * N for _ in range(N)]
    if solve_n_queens_helper(board, 0):
        print_board(board)
    else:
        print("No solution")

if __name__ == "__main__":
    main()
